File: src/presentation/shipsState.ts
import { saveSystem } from "./gameState.ts";
import { loadShipsContent } from "../ships/loadShipsContent.ts";
import type { LoadedShipsContent } from "../ships/loadShipsContent.ts";
import { refreshShipyardPool } from "../ships/refreshShipyardPool.ts";
import { refreshScannerPool } from "../ships/refreshScannerPool.ts";
import type { Ship } from "../data/types/ship.ts";
import type { ShipyardPool } from "../data/types/shipyardPool.ts";
import type { Voyage } from "../data/types/voyage.ts";
import type { ScannerPool } from "../data/types/scannerPool.ts";
import type { Scanner } from "../data/types/scanner.ts";
import type { CombatEncounter } from "../data/types/combatEncounter.ts";
import { SHIPYARD_POOL_REFRESH_INTERVAL_HOURS, SCANNER_POOL_REFRESH_INTERVAL_HOURS } from "../data/constants/shipsAndTravelConfig.ts";
import componentRecipes from "../../content/componentRecipes.json" with { type: "json" };

const MS_PER_HOUR = 60 * 60 * 1000;

// Agent 22's own cross-scene state, same pattern crewState.ts/tradingState.ts
// already established. The shipyard pool is keyed per planet (below) --
// planet-aware fix, see ShipyardScene.ts -- generated on demand for
// whichever planet is actually requested, not just startingPlanet. The
// owned ship roster and active voyages aren't planet-scoped, since a ship
// (once purchased) and its voyages belong to the player, not to any one
// planet.

// Renamed from "profitable:shipyardPool": that key held one ShipyardPool
// total, always for startingPlanet. This holds a per-planet map instead,
// so an old save's single pool is simply regenerated fresh under the new
// key rather than misread as a map.
const SHIPYARD_POOLS_SAVE_KEY = "profitable:shipyardPoolsByPlanet";
const SHIP_ROSTER_SAVE_KEY = "profitable:shipRoster";
const VOYAGES_SAVE_KEY = "profitable:voyages";

// Mirrors tradingState.ts's getTradingContent() exact shape: statically
// imports the one Phase 5 content file loadShipsContent() needs and hands
// it straight through, cached after the first call.
let cachedShipsContent: LoadedShipsContent | null = null;
export function getShipsContent(): LoadedShipsContent {
  if (!cachedShipsContent) {
    cachedShipsContent = loadShipsContent({ componentRecipes });
  }
  return cachedShipsContent;
}

let shipyardPoolsByPlanet: Record<string, ShipyardPool> =
  (saveSystem.load(SHIPYARD_POOLS_SAVE_KEY) as Record<string, ShipyardPool> | null) ?? {};
let roster: Ship[] = (saveSystem.load(SHIP_ROSTER_SAVE_KEY) as Ship[] | null) ?? [];
let voyages: Voyage[] = (saveSystem.load(VOYAGES_SAVE_KEY) as Voyage[] | null) ?? [];

// Generates a planet's pool on first request and persists it, the same
// "load or create, then cache" behavior the old single-pool
// getShipyardPool() had -- just keyed per planet now instead of assuming
// startingPlanet. Re-rolls a fresh pool once
// SHIPYARD_POOL_REFRESH_INTERVAL_HOURS has elapsed since the stored
// pool's own lastRefreshedAt -- a real, previously-missing check (the
// constant existed and was tunable, but nothing ever compared elapsed
// time against it, so a planet's pool never refreshed for the rest of a
// session once generated, despite travel.md's own documentation claiming
// it did). `now` is injectable, same pattern every other time-aware
// function here uses.
export function getShipyardPool(planetId: string, now: number = Date.now()): ShipyardPool {
  const stored = shipyardPoolsByPlanet[planetId];
  if (stored && (now - stored.lastRefreshedAt) / MS_PER_HOUR < SHIPYARD_POOL_REFRESH_INTERVAL_HOURS) {
    return stored;
  }
  const pool = refreshShipyardPool(planetId, undefined, now);
  setShipyardPool(planetId, pool);
  return pool;
}

export function setShipyardPool(planetId: string, next: ShipyardPool): void {
  shipyardPoolsByPlanet = { ...shipyardPoolsByPlanet, [planetId]: next };
  saveSystem.save(SHIPYARD_POOLS_SAVE_KEY, shipyardPoolsByPlanet);
}

export function getShipRoster(): Ship[] {
  return roster;
}

export function setShipRoster(next: Ship[]): void {
  roster = next;
  saveSystem.save(SHIP_ROSTER_SAVE_KEY, roster);
}

export function addShip(ship: Ship): void {
  setShipRoster([...roster, ship]);
}

export function replaceShip(next: Ship): void {
  setShipRoster(roster.map((ship) => (ship.id === next.id ? next : ship)));
}

export function getVoyages(): Voyage[] {
  return voyages;
}

export function setVoyages(next: Voyage[]): void {
  voyages = next;
  saveSystem.save(VOYAGES_SAVE_KEY, voyages);
}

export function addVoyage(voyage: Voyage): void {
  setVoyages([...voyages, voyage]);
}

export function removeVoyage(id: string): void {
  setVoyages(voyages.filter((voyage) => voyage.id !== id));
}

// Scanner/Probe amendment: same cross-scene state pattern as the shipyard
// pool/roster above, applied to scanners' own separate pool (ScannerPool,
// not merged into ShipyardPool, per the Scanner GDD §2.2's explicit
// call-out). Keyed per planet (below), same planet-aware fix as the
// shipyard pool -- see ShipyardScene.ts. Owned scanners aren't
// planet-scoped, same as the ship roster -- once purchased, a scanner
// belongs to the player, usable from wherever their ship is currently
// docked.

// Renamed from "profitable:scannerPool": that key held one ScannerPool
// total, always for startingPlanet. This holds a per-planet map instead.
const SCANNER_POOLS_SAVE_KEY = "profitable:scannerPoolsByPlanet";
const SCANNER_ROSTER_SAVE_KEY = "profitable:scannerRoster";

let scannerPoolsByPlanet: Record<string, ScannerPool> =
  (saveSystem.load(SCANNER_POOLS_SAVE_KEY) as Record<string, ScannerPool> | null) ?? {};
let ownedScanners: Scanner[] = (saveSystem.load(SCANNER_ROSTER_SAVE_KEY) as Scanner[] | null) ?? [];

// Generates a planet's pool on first request and persists it, the same
// "load or create, then cache" behavior the old single-pool
// getScannerPool() had -- just keyed per planet now instead of assuming
// startingPlanet. Re-rolls a fresh pool once
// SCANNER_POOL_REFRESH_INTERVAL_HOURS has elapsed since the stored
// pool's own lastRefreshedAt -- same previously-missing check the
// shipyard/crew pools also lacked; see getShipyardPool()'s own comment.
export function getScannerPool(planetId: string, now: number = Date.now()): ScannerPool {
  const stored = scannerPoolsByPlanet[planetId];
  if (stored && (now - stored.lastRefreshedAt) / MS_PER_HOUR < SCANNER_POOL_REFRESH_INTERVAL_HOURS) {
    return stored;
  }
  const pool = refreshScannerPool(planetId, undefined, now);
  setScannerPool(planetId, pool);
  return pool;
}

export function setScannerPool(planetId: string, next: ScannerPool): void {
  scannerPoolsByPlanet = { ...scannerPoolsByPlanet, [planetId]: next };
  saveSystem.save(SCANNER_POOLS_SAVE_KEY, scannerPoolsByPlanet);
}

export function getOwnedScanners(): Scanner[] {
  return ownedScanners;
}

export function setOwnedScanners(next: Scanner[]): void {
  ownedScanners = next;
  saveSystem.save(SCANNER_ROSTER_SAVE_KEY, ownedScanners);
}

export function addScanner(scanner: Scanner): void {
  setOwnedScanners([...ownedScanners, scanner]);
}

// Combat amendment (Agent 22): a pending CombatEncounter waiting on the
// player's attack/flee choice, paired with a snapshot of the Voyage it was
// detected on. Necessary completion: CombatEncounter itself only carries a
// bare `voyageId` (Agent 1's own contract), but resolveCombatChoice() needs
// the actual origin/destination/cargo to build a retreat voyage -- and by
// the time a combat encounter surfaces here, resolveArrival() has already
// removed the original Voyage from `voyages` (arrival processing completes
// in full regardless of a pending combat, per Agent 20's own contract).
// The snapshot kept here is presentation-only state, not a second copy of
// anything Core persists.
export interface PendingCombat {
  encounter: CombatEncounter;
  voyage: Voyage;
}

const PENDING_COMBATS_SAVE_KEY = "profitable:pendingCombats";
let pendingCombats: PendingCombat[] = (saveSystem.load(PENDING_COMBATS_SAVE_KEY) as PendingCombat[] | null) ?? [];

export function getPendingCombats(): PendingCombat[] {
  return pendingCombats;
}

export function setPendingCombats(next: PendingCombat[]): void {
  pendingCombats = next;
  saveSystem.save(PENDING_COMBATS_SAVE_KEY, pendingCombats);
}

export function addPendingCombat(pending: PendingCombat): void {
  setPendingCombats([...pendingCombats, pending]);
}

export function removePendingCombat(combatEncounterId: string): void {
  setPendingCombats(pendingCombats.filter((pending) => pending.encounter.id !== combatEncounterId));
}

File: unity/ProfitableUnity/Assets/Scripts/Content/ShipsState.cs
#nullable enable
using System.Collections.Generic;
using System.Linq;
using Profitable.Core.Constants;
using Profitable.Core.Schema;
using Profitable.Core.Simulation;

namespace Profitable.Unity.Content
{
    // Migration Phase 2 Sub-Phase D (Ships & Travel) Presentation/
    // Integration -- docs/agents/agent-56-unity-ships-travel-presentation.md.
    // Parallels CrewState.cs/MarketState.cs's own lazy-static shape.
    //
    // Session-only, in-memory, no persistence -- same deliberate scope
    // limit every prior *State.cs class already draws. Shares
    // MarketState.Wallet for ship purchases/refuels, same reasoning
    // CrewState already established for hiring/upkeep/capacity.
    public static class ShipsState
    {
        // Migration Phase 2 Sub-Phase F (Encounters & Combat) addition --
        // docs/agents/agent-63-unity-encounters-combat-presentation.md.
        // A detected-but-unresolved CombatEncounter needs to remember
        // which ship it belongs to -- the Voyage that produced it is
        // discarded once ResolveArrival finishes (ActiveVoyage is cleared
        // immediately), so the encounter alone isn't enough to resolve a
        // later player choice against.
        public sealed class PendingCombatEntry
        {
            public CombatEncounter Encounter { get; set; } = new();
            public string ShipId { get; set; } = string.Empty;

            // The voyage that produced this encounter, kept only so a
            // later ResolveCombatChoice call can rebuild the origin/
            // current planets and cargo a retreat voyage needs -- not
            // read for any other purpose.
            public Voyage Voyage { get; set; } = new();
        }

        private static List<Ship>? _ownedShips;
        private static ShipyardPool? _shipyardPool;
        private static Voyage? _activeVoyage;
        private static List<PendingCombatEntry>? _pendingCombats;
        private static ScannerPool? _scannerPool;
        private static List<Scanner>? _ownedScanners;

        public static List<Ship> OwnedShips => _ownedShips ??= new List<Ship>();
        public static List<PendingCombatEntry> PendingCombats => _pendingCombats ??= new List<PendingCombatEntry>();
        public static List<Scanner> OwnedScanners => _ownedScanners ??= new List<Scanner>();

        public static void ReplaceShip(Ship updated)
        {
            var index = OwnedShips.FindIndex(s => s.Id == updated.Id);
            if (index >= 0) OwnedShips[index] = updated;
        }

        // Same "compare elapsed time since LastRefreshedAt against the
        // refresh interval, re-roll when stale" pattern CrewState already
        // established for its own pool -- the dead-pool-refresh fix this
        // session's earlier TypeScript work applied to
        // getShipyardPool()/getScannerPool() too.
        public static ShipyardPool GetOrRefreshShipyardPool(long nowMs)
        {
            if (_shipyardPool is null || IsStale(_shipyardPool.LastRefreshedAt, nowMs, ShipsAndTravelConfig.ShipyardPoolRefreshIntervalHours))
            {
                _shipyardPool = ShipyardPoolRefresher.RefreshShipyardPool(GalaxyState.StartingPlanet.Id, seed: null, nowMs);
            }
            return _shipyardPool;
        }

        public static void SetShipyardPool(ShipyardPool pool) => _shipyardPool = pool;

        // Scanner UI gap closed (2026-08-04) -- mirrors
        // GetOrRefreshShipyardPool's own stale-pool-refresh pattern
        // exactly, for the scanner pool this Unity build never surfaced
        // before now.
        public static ScannerPool GetOrRefreshScannerPool(long nowMs)
        {
            if (_scannerPool is null || IsStale(_scannerPool.LastRefreshedAt, nowMs, ShipsAndTravelConfig.ScannerPoolRefreshIntervalHours))
            {
                _scannerPool = ScannerPoolRefresher.RefreshScannerPool(GalaxyState.StartingPlanet.Id, seed: null, nowMs);
            }
            return _scannerPool;
        }

        public static void SetScannerPool(ScannerPool pool) => _scannerPool = pool;

        public static Voyage? ActiveVoyage => _activeVoyage;
        public static void SetActiveVoyage(Voyage? voyage) => _activeVoyage = voyage;

        private static bool IsStale(long lastRefreshedAt, long nowMs, double refreshIntervalHours)
        {
            var elapsedHours = (nowMs - lastRefreshedAt) / (60.0 * 60.0 * 1000.0);
            return elapsedHours >= refreshIntervalHours;
        }

        // Mirrors every other *State.cs's own ResetForTests() hook.
        public static void ResetForTests()
        {
            _ownedShips = null;
            _shipyardPool = null;
            _activeVoyage = null;
            _pendingCombats = null;
            _scannerPool = null;
            _ownedScanners = null;
        }
    }
}
